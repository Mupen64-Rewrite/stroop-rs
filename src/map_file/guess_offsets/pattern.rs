pub struct Pattern {
    pattern: &'static [Option<u8>],
    offset: usize,
}

impl Pattern {
    pub fn matches(&self, mut data: &[u8]) -> Option<usize> {
        // we don't let pattern to be constructed by the user, so head is always Some
        let head = self.pattern[0].unwrap();
        let mut data_offset = 0;

        loop {
            // advance data slice to head
            data = match data.iter().position(|&f| f == head) {
                Some(index) => {
                    data_offset += index;
                    &data[index..]
                }
                None => return None,
            };

            // not enough bytes?
            if data.len() < self.pattern.len() {
                return None;
            }

            // check if pattern matches
            for pattern in self.pattern[1..].iter() {
                if let Some(pattern) = pattern {
                    if data[0] != *pattern {
                        data_offset += 1;
                        data = &data[1..];
                        continue;
                    }
                }
            }

            return Some(data_offset + self.offset);
        }
    }
}
