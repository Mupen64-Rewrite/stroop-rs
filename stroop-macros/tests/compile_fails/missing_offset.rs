use stroop_macros::pattern;

fn main() {
    pattern!();
}
