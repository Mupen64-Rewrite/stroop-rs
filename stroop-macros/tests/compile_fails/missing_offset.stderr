error: expected offset as isize
 --> tests/compile_fails/missing_offset.rs:4:5
  |
4 |     pattern!();
  |     ^^^^^^^^^^
  |
  = note: this error originates in the macro `pattern` (in Nightly builds, run with -Z macro-backtrace for more info)
